File: Java/EuclidAlgorithm/src/main/java/org/example/euclid/Euclid.java
package org.example.euclid;

/**
 * Euclid Algorithm
 * It calculates the GCD(A,B) = Greatest Common Divisor between two numbers A and B
 */
public class Euclid {

    /**
     * Implementation using recursion<BR>
     *     22/6 = 3 rest of 4
     *     6/4 = 1 rest of 2
     *     4/2 = 2 rest of 0
     * @param number
     * @param divisor
     * @return
     */
    public int gcd(int number, int divisor) {
        int remaining = (number % divisor);

        if (remaining != 0) {
            return gcd(divisor, remaining);
        } else {
            return divisor;
        }
    }

    /**
     * Implementation without recursion
     *      22/6 = 3 rest of 4
     *      6/4 = 1 rest of 2
     *      4/2 = 2 rest of 0
     *
     *      number/temp = result rest of divisor
     * @param number
     * @param divisor
     * @return
     */
    public int gcd2(int number, int divisor) {
        while (divisor != 0) {
            int temp = divisor;
            divisor = (number % divisor);
            number = temp;
        }

        return number;
    }
}

File: Kotlin/EuclidAlgorithm/src/main/kotlin/euclid/Euclid.kt
package euclid

/**
 * Euclid Algorithm
 * It calculates the GCD(A,B) = Greatest Common Divisor between two numbers A and B
 */
class Euclid {

    /**
     * Implementation using recursion<BR>
     *     22/6 = 3 rest of 4
     *     6/4 = 1 rest of 2
     *     4/2 = 2 rest of 0
     * @param number
     * @param divisor
     * @return
     */
    fun gcd(number: Int, divisor: Int): Int {
        val remaining = (number % divisor);
        return if (remaining != 0) {
            gcd(divisor, remaining)
        } else {
            divisor
        }
    }

    /**
     * Implementation without recursion
     *      22/6 = 3 rest of 4
     *      6/4 = 1 rest of 2
     *      4/2 = 2 rest of 0
     *
     *      number/temp = result rest of divisor
     * @param number
     * @param divisor
     * @return
     */
    fun gcd2(_number: Int, _divisor: Int): Int {
        var number = _number
        var divisor = _divisor
        while (divisor != 0) {
            val temp: Int = divisor
            divisor = number % divisor
            number = temp
        }
        return number
    }


}